frame/composable-traits/src/dex.rs

```diff
@@ -33,20 +33,34 @@ pub trait Amm {
 	fn lp_token(pool_id: Self::PoolId) -> Result<Self::AssetId, DispatchError>;
 
 	/// Returns the amount of base & quote asset redeemable for given amount of lp token.
-	fn redeemable_assets_for_given_lp_tokens(
+	fn redeemable_assets_for_lp_tokens(
 		pool_id: Self::PoolId,
 		lp_amount: Self::Balance,
+		min_expected_amounts: BTreeMap<Self::AssetId, Self::Balance>,
 	) -> Result<RedeemableAssets<Self::AssetId, Self::Balance>, DispatchError>
 	where
 		Self::AssetId: sp_std::cmp::Ord;
 
-	/// Returns the amount of LP tokens that would be recieved by adding the given amounts of base
-	/// and quote.
-	fn amount_of_lp_token_for_added_liquidity(
+	/// Simulate add_liquidity computations, on success returns the amount of LP tokens
+	/// that would be recieved by adding the given amounts of base and quote.
+	fn simulate_add_liquidity(
+		who: &Self::AccountId,
 		pool_id: Self::PoolId,
-		base_amount: Self::Balance,
-		quote_amount: Self::Balance,
-	) -> Result<Self::Balance, DispatchError>;
+		amounts: BTreeMap<Self::AssetId, Self::Balance>,
+	) -> Result<Self::Balance, DispatchError>
+	where
+		Self::AssetId: sp_std::cmp::Ord;
+
+	/// Simulate remove_liquidity computations, on success returns the amount of base/quote assets
+	/// that would be recieved by removing the given amounts of lp tokens.
+	fn simulate_remove_liquidity(
```

Comment:
Why do we add another function for that, RPC are readonly so you can directly use the usual add/remove liquidity function from a RPC to dry run?

Reply:
I was not aware about that. So I added a similar function which just check if balances are enough on all sides.

Reply:
@hussein-aitlahcen so if I call a function which has #transactional from RPC , it will not change any thing in storage ?

Reply:
I also would like to know a definitive answer to this question :) , what happens when an RPC call is made that updates storage? Does it avoid any storage updates if there is no explicit initiation of an extrinsic?

Reply:
It's a readonly operation that is not going to do anthing to the storage. So we are free to declare RPC that call functions called by extrinsics safely.

Reply:
We need to be careful though as a node can be ddos, but if it's a RPC only node we can probably scale it. Computing the liquidity is quite fast

Reply:
By the way, you should check out RPCs:
```
    system_dryRun
    system_dryRunAt
```

Reply:
@saadjhk could you please verify https://polkadot.js.org/docs/substrate/rpc/#dryrunextrinsic-bytes-at-blockhash-applyextrinsicresult if that can satisfy your needs?

Reply:
@hussein-aitlahcen those RPCs require signed extrinsics, FE will have to prompt users for signatures for each dryrun